sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java

```diff
@@ -512,6 +581,18 @@ String retrieveErrorDetails(Iterable<AppendRowsContext> failedContext) {
             .map(StackTraceElement::toString)
             .collect(Collectors.joining("\n"));
       }
+
+      void postFlush() {
+        // If we got a response indicating an updated schema, recreate the client.
+        if (updatedTableSchema != null
+            && this.appendClientInfo != null
+            && this.appendClientInfo.hasSchemaChanged(updatedTableSchema)) {
+          invalidateWriteStream();
+          appendClientInfo =
+              Preconditions.checkStateNotNull(getAppendClientInfo(false, updatedTableSchema));
+          updatedTableSchema = null;
```

Comment:
I am wondering if there will be races regarding this updatedSchema?

Reply:
what race do you envision?

Reply:
Some responses are coming back and the updatedTableSchema is being updated L568, which in race with the postFlush here.

Reply:
postFlush is called only after all futures have completed (in flushAll), so we would not expect any more callbacks. Also note that RetryManager calls these response callbacks in the primary thread (RetryManager.await() calls the callbacks), so the callbacks here are not being called asynchronously.